src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c

```diff
@@ -10,9 +10,199 @@
+        tl_error(lib, "failed to get cm event, errno %d", errno);
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    rdma_ack_cm_event(revent);
+    ctx->ctx = ctx->id->verbs;
+    ctx->pd  = ibv_alloc_pd(ctx->ctx);
+    if (!ctx->pd) {
+        tl_error(lib, "failed to allocate pd");
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    ib           = strdup(ctx->devname);
+    ucs_string_split(ib, ":", 2, &ib_name, &port);
+    ctx->ib_port = atoi(port);
+
+    /* Determine MTU */
+    if (ibv_query_port(ctx->ctx, ctx->ib_port, &port_attr)) {
+        tl_error(lib, "couldn't query port in ctx create, errno %d", errno);
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    if (port_attr.max_mtu == IBV_MTU_256)
+        max_mtu = 256;
+    if (port_attr.max_mtu == IBV_MTU_512)
+        max_mtu = 512;
+    if (port_attr.max_mtu == IBV_MTU_1024)
+        max_mtu = 1024;
+    if (port_attr.max_mtu == IBV_MTU_2048)
+        max_mtu = 2048;
+    if (port_attr.max_mtu == IBV_MTU_4096)
+        max_mtu = 4096;
+
+    if (port_attr.active_mtu == IBV_MTU_256)
+        active_mtu = 256;
+    if (port_attr.active_mtu == IBV_MTU_512)
+        active_mtu = 512;
+    if (port_attr.active_mtu == IBV_MTU_1024)
+        active_mtu = 1024;
+    if (port_attr.active_mtu == IBV_MTU_2048)
+        active_mtu = 2048;
+    if (port_attr.active_mtu == IBV_MTU_4096)
+        active_mtu = 4096;
+
+    ctx->mtu = active_mtu;
+
+    if (port_attr.max_mtu < port_attr.active_mtu) {
+        tl_debug(ctx->lib, "port active MTU (%d) is smaller than port max MTU (%d)",
+                    active_mtu, max_mtu);
+    }
+    if (ibv_query_device(ctx->ctx, &device_attr)) {
+        tl_error(lib, "failed to query device in ctx create, errno %d", errno);
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    tl_debug(ctx->lib, "MTU %d, MAX QP WR: %d, max sqr_wr: %d, max cq: %d, max cqe: %d",
+                ctx->mtu, device_attr.max_qp_wr, device_attr.max_srq_wr,
+                device_attr.max_cq, device_attr.max_cqe);
+
+    ctx->max_qp_wr = device_attr.max_qp_wr;
+    {
+        status = ucc_mpool_init(&ctx->compl_objects_mp, 0, sizeof(ucc_tl_mlx5_mcast_p2p_completion_obj_t), 0,
+                                UCC_CACHE_LINE_SIZE, 8, UINT_MAX,
+                                &ucc_coll_task_mpool_ops,
+                                UCC_THREAD_SINGLE,
+                                "ucc_tl_mlx5_mcast_p2p_completion_obj_t");
+        if (ucc_unlikely(UCC_OK != status)) {
+            tl_error(lib, "failed to initialize compl_objects_mp mpool");
+            status = UCC_ERR_NO_MEMORY;
+            goto error;
+        }
+    }
+
+    ctx->rcache = NULL;
+    if (UCC_OK != ucc_tl_mlx5_mcast_setup_rcache(ctx)) {
+        tl_error(lib, "failed to setup rcache");
+        status = UCC_ERR_NO_RESOURCE;
+        goto error;
+    }
+
+    tl_debug(ctx->lib, "tl mcast setup complete: ctx %p", ctx);
```

Comment:
"tl mcast" is not needed, it's added by tl_debug.
"setup complete" is ambiguous, "context setup complete" may be better